File: java/src/main/java/structural/Bridge.java
package structural;

// Implementation
interface Renderer {

    void render(String shapeName);
}

class OpenGLRenderer implements Renderer {

    @Override
    public void render(String shapeName) {
        System.out.println("Rendering " + shapeName + " with OpenGL");
    }
}

class DirectXRenderer implements Renderer {

    @Override
    public void render(String shapeName) {
        System.out.println("Rendering " + shapeName + " with DirectX");
    }
}

// Abstraction
abstract class Shape {

    protected Renderer renderer;

    public Shape(Renderer renderer) {
        this.renderer = renderer;
    }

    public abstract void draw();
}

class Circle extends Shape {

    public Circle(Renderer renderer) {
        super(renderer);
    }

    @Override
    public void draw() {
        renderer.render("Circle");
    }
}

class Square extends Shape {

    public Square(Renderer renderer) {
        super(renderer);
    }

    @Override
    public void draw() {
        renderer.render("Square");
    }
}

@SuppressWarnings("unused")
final class BridgeExample {

    private BridgeExample() {
    }

    public static void example() {
        Shape circle = new Circle(new OpenGLRenderer());
        Shape square = new Square(new DirectXRenderer());

        circle.draw(); // Rendering Circle with OpenGL
        square.draw(); // Rendering Square with DirectX
    }
}

File: rust/src/structural/bridge.rs
// Implementation side (Renderer)
trait Renderer {
    fn render(&self, shape_name: &str);
}

struct OpenGLRenderer;
struct DirectXRenderer;

impl Renderer for OpenGLRenderer {
    fn render(&self, shape_name: &str) {
        println!("Rendering {} using OpenGL", shape_name);
    }
}

impl Renderer for DirectXRenderer {
    fn render(&self, shape_name: &str) {
        println!("Rendering {} using DirectX", shape_name);
    }
}

// Abstraction side (Shape)
trait Shape {
    fn draw(&self);
}

// Concrete shape Circle
struct Circle {
    renderer: Box<dyn Renderer>,
}

impl Circle {
    fn new(renderer: Box<dyn Renderer>) -> Self {
        Circle { renderer }
    }
}

impl Shape for Circle {
    fn draw(&self) {
        self.renderer.render("Circle");
    }
}

// Concrete shape Square
struct Square {
    renderer: Box<dyn Renderer>,
}

impl Square {
    fn new(renderer: Box<dyn Renderer>) -> Self {
        Square { renderer }
    }
}

impl Shape for Square {
    fn draw(&self) {
        self.renderer.render("Square");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example() {
        let opengl = Box::new(OpenGLRenderer);
        let directx = Box::new(DirectXRenderer);

        let circle = Circle::new(opengl);
        let square = Square::new(directx);

        circle.draw(); // Rendering Circle using OpenGL
        square.draw(); // Rendering Square using DirectX
    }
}
